Weights in alpha for FocalLoss

ytl0623 opened the pull request:
Fixes #8601

### Description

Support alpha as a list, tuple, or tensor of floats, in addition to the existing scalar support.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

ytl0623 committed:
Weights in alpha for FocalLoss

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 committed:
fix Local variable lpha_arg is assigned to but never used

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 committed:
fix mypy error

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 committed:
fix undefined type error

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 committed:
fix alpha type bugs

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 committed:
fix alpha type bugs

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 commented:
Hi @ericspod @KumoLiu @Nic-Ma ,

sorry to bother u. the CI is failing with `[Errno 28] No space left on device` during the build-docs and packaging steps. This seems to be a server-side issue. Could u plz re-trigger the workflows?

Thanks in advance!

ytl0623 committed:
Merge branch 'dev' into fix-issue-8601

ericspod commented on monai/losses/focal_loss.py:
```diff
@@ -156,13 +164,19 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+
+        alpha_arg: float | torch.Tensor | None = self.alpha
+        if isinstance(alpha_arg, torch.Tensor):
+            alpha_arg = alpha_arg.to(input.device)
```
If this necessary here if `sigmoid_focal_loss` and `softmax_focal_loss` do this conversion as well?

ytl0623 replied:
You are right. This block is redundant, so I will remove it.

ericspod commented:
Hi @ytl0623 thanks for this, we should have the CICD issues sorted now. It looks good to me though I think we need to add tests for the class with a sequence `alpha` value to ensure this works correctly. Please add this then update the branch to incorporate the CICD fixes.

ytl0623 committed:
Merge branch 'Project-MONAI:dev' into fix-issue-8601

ytl0623 committed:
add test case for alpha as a sequence

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 commented:
> Hi @ytl0623 thanks for this, we should have the CICD issues sorted now. It looks good to me though I think we need to add tests for the class with a sequence `alpha` value to ensure this works correctly. Please add this then update the branch to incorporate the CICD fixes.

Hi @ericspod, thanks for the feedback. I have removed the redundant conversion logic and added a new test case. It should be ready for another look. Thanks in advance!

ytl0623 committed:
Merge branch 'dev' into fix-issue-8601

ericspod commented on tests/losses/test_focal_loss.py:
```diff
@@ -374,6 +374,47 @@ def test_script(self):
+        batch_size = 2
+        spatial_dims = (4, 4)
+
+        devices = ["cpu", "cuda"] if torch.cuda.is_available() else ["cpu"]
```
We have [this definition](https://github.com/Project-MONAI/MONAI/blob/01711cf7e3af991aabc92cc3fdf4c0b65949e65e/tests/test_utils.py#L874C1-L874C14) to use in place of this line if you want. Another approach is to have a `device` argument and use the `parameterized.expand` decorator [like here](https://github.com/Project-MONAI/MONAI/blob/01711cf7e3af991aabc92cc3fdf4c0b65949e65e/tests/data/meta_tensor/test_meta_tensor.py#L201).

ericspod commented on tests/losses/test_focal_loss.py:
```diff
@@ -374,6 +374,47 @@ def test_script(self):
             test_input = torch.ones(2, 2, 8, 8)
             test_script_save(loss, test_input, test_input)
 
+    def test_alpha_sequence_broadcasting(self):
```
Here I think we do need another test, or use `parameterized.expand`, to test with and without the background.

ytl0623 replied:
Done.

ericspod commented:
Just about there, let's add a test for background=False and the stacklevel change for the warning and we'll be good to go.

ytl0623 committed:
another test  to test without the background

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 commented:
Thank you for the feedback. I have refactored the tests to use `@parameterized.expand`.

ytl0623 committed:
Merge branch 'Project-MONAI:dev' into fix-issue-8601

ytl0623 committed:
fix bug: scalar alpha still passed despite ignored warning

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 committed:
Merge branch 'dev' into fix-issue-8601

Signed-off-by: ytl0623 <david89062388@gmail.com>

ytl0623 committed:
fix: When alpha is a sequence, each alpha[c] should be interpreted as the weight for positive samples of class c. Negative samples should have a default weight of 1.0

Signed-off-by: ytl0623 <david89062388@gmail.com>

ericspod reviewed (approved):
Looks good now!

ytl0623 committed:
Merge branch 'dev' into fix-issue-8601

ytl0623 committed:
Merge branch 'dev' into fix-issue-8601

ytl0623 committed:
Merge branch 'dev' into fix-issue-8601

ytl0623 committed:
Merge branch 'dev' into fix-issue-8601

ericspod committed:
Merge branch 'dev' into fix-issue-8601
